src/gui/SampleThumbnail.cpp

```diff
@@ -0,0 +1,216 @@
+	s_sampleThumbnailCacheMap.insert(std::make_pair(name, m_thumbnailCache));
+	return false;
+}
+
+/* DEPRECATED; functionality is kept for testing conveniences */
+void SampleThumbnail::cleanUpGlobalThumbnailMap()
+{
+	auto map = s_sampleThumbnailCacheMap.begin();
+	while (map != s_sampleThumbnailCacheMap.end())
+	{
+		// All sample thumbnails are destroyed, a.k.a sample goes out of use
+		if (map->second.use_count() == 1)
+		{
+			s_sampleThumbnailCacheMap.erase(map);
+			map = s_sampleThumbnailCacheMap.begin();
+			continue;
+		}
+
+		map++;
+	}
+}
+
+SampleThumbnail::Thumbnail SampleThumbnail::generate(const size_t thumbnailSize, const SampleFrame* buffer, const size_t size)
+{
+	const auto sampleChunk = (size + thumbnailSize) / thumbnailSize;
+	auto thumbnail = Thumbnail(thumbnailSize);
+
+	for (auto tIndex = size_t{0}; tIndex < thumbnailSize; tIndex++)
+	{
+		const auto
+		sampleIndex = tIndex * size / thumbnailSize,
+		sampleChunkBound = std::min(sampleIndex + sampleChunk, size);
+
+		thumbnail[tIndex] = std::accumulate(buffer + sampleIndex, buffer + sampleChunkBound, Bit{});
+	}
+
+	return thumbnail;
+}
+
+SampleThumbnail::SampleThumbnail(const Sample& inputSample)
+{
+	if (selectFromGlobalThumbnailMap(inputSample)) { return; }
+
+	cleanUpGlobalThumbnailMap();
+
+	const auto sampleBufferSize 	= inputSample.sampleSize();
+	const auto thumbnailSizeDivisor = size_t{10};
+	const auto firstThumbnailSize 	= std::max<size_t>(sampleBufferSize / thumbnailSizeDivisor, 1);
+
+	const auto firstThumbnail 		= generate(firstThumbnailSize, inputSample.data(), sampleBufferSize);
+
+	m_thumbnailCache->push_back(firstThumbnail);
+
+	// Generate the remaining thumbnails using the first one, each one's
+	// size is the size of the previous one divided by the thumbnail size divisor.
+	auto thumbnailSize = size_t{firstThumbnailSize / thumbnailSizeDivisor};
+	while(thumbnailSize >= 1)
+	{
+		const auto& biggerThumbnail = m_thumbnailCache->back();
+
+		auto thumbnail = Thumbnail(thumbnailSize);
+
+		for (auto
+			smallIndex = size_t{0};
+			smallIndex < thumbnail.size();
+			smallIndex += 1
+		) {
+			auto bigIndex 	= smallIndex * thumbnailSizeDivisor;
+			auto bigItStart = biggerThumbnail.cbegin() + bigIndex;
+			auto bigItEnd 	= std::min(bigItStart + thumbnailSizeDivisor, biggerThumbnail.end());
+
+			thumbnail[smallIndex] = std::accumulate(bigItStart, bigItEnd, Bit{});
+		}
+
+		m_thumbnailCache->push_back(thumbnail);
+		thumbnailSize /= thumbnailSizeDivisor;
+	}
+}
+
+void SampleThumbnail::draw(QPainter& painter, const Bit& bit, float lineX, int centerY,
+	float scalingFactor, const QColor& color, const QColor& innerColor)
+{
+	const auto lengthY1 = bit.max * scalingFactor;
+	const auto lengthY2 = bit.min * scalingFactor;
+
+	const auto lineY1 = centerY - lengthY1;
+	const auto lineY2 = centerY - lengthY2;
+
+	const auto innerLineY1 = centerY - bit.max*0.6 * scalingFactor;
+	const auto innerLineY2 = centerY - bit.min*0.6 * scalingFactor;
```

Comment:
What's with the `0.6` stuff? Shouldn't you be using the RMS value here too?

Reply:
Oh wait you removed RMS (probably could've been in another PR to make sure other people approve of those changes and to avoid possible backlash)

I think you can remove all of the inner line stuff since RMS is gone now.

Reply:
The RMS value was problematic while optimizing the code because I can't get the correct value using accumulate so I replaced it with a placeholder. I still think RMS doesn't have a lot of significance in the waveform view though, but I'll try to add it back in the next commits

Reply:
Noo, please don't interpret my suggestions as commands or anything. I am okay with the RMS removal too, I was just taking possible backlash into consideration.

If you want to add it back in, adding the RMS would probably require storing a squared sum within the bits. Then you can just calculate the RMS within the draw function.

Reply:
(Github wasn't updating so I didn't see your 2nd message)

We still need to know how many bits were merged to get the correct mean value. I still don't know how to solve that yet, so I'll leave the RMS out for now.

Reply:
Yeah, it might not be worth it since you'd be carrying around a bit more in each bit.

If any major backlash (very unlikely) it can be added back in.

Reply:
I still think you can remove the inner line stuff though. This was for the RMS calculations, but now they are no longer needed.

Reply:
I think it doesn't hurt to keep the inner lines. Without them the waveform looks plain and kind of bland (at least in the song editor)

Reply:
But then we aren't drawing anything of significance. That feels very wrong 🙁

Reply:
Yeah you're right. I'll do ahead and remove that as well

Reply:
The waveform looking bland was the kind of backlash I was referring to. It won't bother me but when someone notices something has visually changed, it may raise eyebrows, even if it wasn't actively taken into consideration before.

Other DAWs do not draw RMS and the waveform looks fine. I think the root problem is how we are drawing sample clips. I'm just not sure what would need to change in that case.

Reply:
I'll see if I can go about adding it back in while making sure it still works when merging bits.

Reply:
Maybe we can ask other devs (probably users as well) for opinions about this change and do what's preferred

Reply:
Okay I've asked: it doesn't seem important to people at all.

Reply:
Looks like we can safely remove it then